spirv_cross/src/glsl.rs

```diff
@@ -135,11 +141,39 @@ impl spirv::Compile<Target> for spirv::Ast<Target> {
 
     /// Generate GLSL shader from the AST.
     fn compile(&mut self) -> Result<String, ErrorCode> {
+        self.build_combined_image_samplers()?;
+        self.compiler.compile()
+    }
+}
+
+impl spirv::Ast<Target> {
+    pub fn build_combined_image_samplers(&mut self) -> Result<(), ErrorCode> {
+        unsafe {
+            if !self.compiler.target_data.combined_image_samplers_built {
+                check!(sc_internal_compiler_glsl_build_combined_image_samplers(
+                    self.compiler.sc_compiler
+                ));
+                self.compiler.target_data.combined_image_samplers_built = true
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn get_combined_image_samplers(&mut self) -> Result<&[ScCombinedImageSampler], ErrorCode> {
+        self.build_combined_image_samplers()?;
```

Comment:
We shouldn't be returning `Sc*` in the public API. Let's copy these over to our own `Vec<CombinedImageSampler>` and return that instead (like some of the other functions do), and avoid sharing memory with the C++ values. The reason I prefer this is to control the lifetime of these structs on the Rust side – for example, we have no way to know if the combined image samplers are rebuilt multiple times and the old one is invalidated.

Reply:
Can't be rebuilt multiple times cause we have a bool to stop that, but I understand what you mean. Fixed.